esp-hal-embassy/src/executor/thread.rs

```diff
@@ -116,6 +116,10 @@ This will use software-interrupt 3 which isn't available for anything else to wa
     }
 
     #[cfg(all(xtensa, low_power_wait))]
+    // This function must be in RAM. Loading parts of it from flash can cause a race
+    // that results in the core not waking up. Placing `wait_impl` in RAM ensures that
+    // it is shorter than the interrupt handler that would clear the interrupt source.
+    #[macros::ram]
     fn wait_impl(cpu: usize) {
```

Comment:
I don't know if this is the same or separate issue but I think this `wait_impl` implementation is incorrect for the S2.

This function uses RSIL and WSR.PS/WAITI to create a critical section, however, the S2 doesn't natively support atomics, so portable-atomic actually [disables and re-enables](https://github.com/taiki-e/portable-atomic/blob/8bd3b5d267a69d37ab31b74ceb845a8345c1eaf5/src/imp/interrupt/xtensa.rs#L28) interrupts to work.... oh, as I write this I realise this is fine. Worse case it is just wasted CPU cycles. The atomic nature of the critical section is unaffected I think.

I'll just leave this comment here for someone else to review and confirm :sweat_smile:

Reply:
I'm not sure of the entire mechanism that causes it, but here's the generated code for S2, you can see there's a single rsil there:

![image](https://github.com/user-attachments/assets/bef87a2f-5240-4f5a-a391-5780280bd2ba)

Wedeclare [+forced-atomics](https://github.com/rust-lang/rust/blob/d9a739363878a06633eb912e44f02a3c5d6ba08c/compiler/rustc_target/src/spec/targets/xtensa_esp32s2_none_elf.rs#L22) for the S2, which causes `cfg_target_has_atomic` to be defined.

portable-atomic has a "has atomic load/store" check that makes it fall back to `core::atomic` where it can.

I'm not sure exactly how the pieces fit together beyond this point.

Reply:
Ah I see, thanks for checking!